### PR TITLE
FIx bug `density_matrix` and torch

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
When using torch returning the density matrix yields wrong format results.

**Description of the Change:**
The Torch `_asarray` function in `default.qubit.torch` is always unrolling the tensor. I created a specific case for density matrix.

**Benefits:**
`qml.density_matrix` works with Torch interface.

**Related GitHub Issues:**
Closes https://github.com/PennyLaneAI/pennylane/issues/2553

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-19818]

File: tests/test_measurements.py [hunk 1]
```diff
@@ -1208,6 +1208,24 @@ def func():
         assert len(obs) == 1
         assert obs[0].return_type is State
 
+    @pytest.mark.torch
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_torch(self, dev_name):
+        """Test that the correct density matrix is returned using torch interface."""
```
Comment (eddddddy) [inline tests/test_measurements.py:1214]: Seems strange to have a single standalone test for PyTorch here. I think maybe `tests/devices/default_qubit_torch.py` would be better.

Also, do we test density matrices for the other interfaces already?

**Review by eddddddy — COMMENTED** [review]
Hi @rmoyard, looks good to me! Just a minor change and a question, then should be good to go.

File: tests/test_measurements.py (hunk 1 above)
Comment (rmoyard) [reply tests/test_measurements.py:1214]: I added more tests here (tf, jax), and it will be the case in the upcoming PRs 👍

File: tests/test_measurements.py [hunk 2]
```diff
@@ -1208,6 +1208,60 @@ def func():
...
+
+        density_mat = func()
+
+        assert np.allclose(
+            np.array([[0.5 + 0.0j, 0.5 + 0.0j], [0.5 + 0.0j, 0.5 + 0.0j]]), density_mat
+        )
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_torch(self, dev_name):
```
Comment (eddddddy) [inline tests/test_measurements.py:1231]:
```suggestion
    def test_correct_density_matrix_jax(self, dev_name):
```

File: tests/test_measurements.py [hunk 3]
```diff
@@ -1208,6 +1208,60 @@ def func():
...
+
+        density_mat = func()
+
+        assert np.allclose(
+            np.array([[0.5 + 0.0j, 0.5 + 0.0j], [0.5 + 0.0j, 0.5 + 0.0j]]), density_mat
+        )
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_torch(self, dev_name):
```
Comment (eddddddy) [inline tests/test_measurements.py:1249]:
```suggestion
    def test_correct_density_matrix_tf(self, dev_name):
```

**Review by eddddddy — APPROVED** [review]

**Review by antalszava — COMMENTED** [review]
@rmoyard this is awesome! :1st_place_medal: :tada: 

There was a [major thing I wondered about related to the change in logic](https://github.com/PennyLaneAI/pennylane/pull/2643#discussion_r886874619), though the fix seems to overall check out. :+1:

File: tests/test_measurements.py [hunk 4]
```diff
@@ -1208,6 +1208,60 @@ def func():
...
         assert obs[0].return_type is State
 
+    @pytest.mark.torch
+    @pytest.mark.parametrize("dev_name", ["default.qubit", "default.mixed"])
+    def test_correct_density_matrix_torch(self, dev_name):
+        """Test that the correct density matrix is returned using torch interface."""
+
+        dev = qml.device(dev_name, wires=2)
+
+        @qml.qnode(dev, interface="torch")
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/test_measurements.py:1218]: I don't think that the issue had anything to do with backprop at the end but I will add a diff method fixture 👍

File: pennylane/devices/default_qubit_torch.py [hunk 5]
```diff
@@ -239,9 +239,12 @@ def _asarray(self, a, dtype=None):
             if not isinstance(a[0], torch.Tensor):
                 res = np.asarray(a)
                 res = torch.from_numpy(res)
+                res = torch.cat([torch.reshape(i, (-1,)) for i in res], dim=0)
+            elif len(a) == 1 and len(a[0].shape) > 1:
+                res = torch.cat(a, dim=0)
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/devices/default_qubit_torch.py:244]: I tried this solution but unfortunately it is still unrolling the density matrix, if I don't use `torch.cat(a, dim=0)` I also run into an error. But I think the logic works well because it is only applied when there is only one measurement and therefore it should be safe. Another idea @antalszava ?

File: tests/test_measurements.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply tests/test_measurements.py:1218]: Added!

File: pennylane/devices/default_qubit_torch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/devices/default_qubit_torch.py:244]: With the previous logic you are doing `[tensor([[0.5000+0.j, 0.5000+0.j],
        [0.5000+0.j, 0.5000+0.j]], dtype=torch.complex128)]` ->
`tensor([0.5000+0.j, 0.5000+0.j, 0.5000+0.j, 0.5000+0.j],
       dtype=torch.complex128)` ->
`tensor([0.5000+0.j, 0.5000+0.j, 0.5000+0.j, 0.5000+0.j],
       dtype=torch.complex128)`

[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/devices/default_qubit_torch.py:244]: with the new logic:
`a=[tensor([[0.5000+0.j, 0.5000+0.j],
        [0.5000+0.j, 0.5000+0.j]], dtype=torch.complex128)]`
->
`res=tensor([[0.5000+0.j, 0.5000+0.j],
        [0.5000+0.j, 0.5000+0.j]], dtype=torch.complex128)`

[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/devices/default_qubit_torch.py:243]: Oh, I see! Then a minor suggestion: how about doing `res=a[0]`? Concatenation seems a bit of an overkill to achieve that.

**Review by antalszava — APPROVED** [review]
Looks good :+1:

File: pennylane/devices/default_qubit_torch.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/devices/default_qubit_torch.py]: I changed it 👍
